keps/sig-node/4383-dra-semantic-model-mode-switching/README.md

```diff
@@ -0,0 +1,898 @@
+parameters" ([KEP
+#4381](https://github.com/kubernetes/enhancements/issues/4381)) to be usable
+with Dynamic Resource Allocation (DRA).
+
+This KEP defines a semantic model that roughly corresponds in functionality to
+what the initial set of DRA drivers for GPUs supported with vendor CRDs.
+
+## Motivation
+
+### Goals
+
+* Enable DRA GPU drivers to support cluster autoscaling.
+
+### Non-Goals
+
+* Provide a model that can handle all future hardware. Other models may be
+  needed for things that don't fit into the schema defined here.
+
+## Proposal
+
+The resource model is that each node contains a certain set of resource
+instances. Each resource instance can be in one of several different, finite
+modes. Each mode has different attributes. Some of these attributes are labels
+that allow selecting a suitable instance, others are counters of things that
+are available in the mode for consumption by different resource claims.
+
+Each instance can be only in one mode at once. It can switch between modes only
+when it is not currently in use.
+
+### Risks and Mitigations
+
+Adding another semantic model doesn't add any additional risks. There is a
+slight performance impact when checking whether the builtin controller of the
+semantic model needs to do something for a claim or NodeResourceCapacity object,
+but that impact should be minimal when the model is not in use.
+
+This resource model is fairly complex. Finding the right instance and mode of
+that instance to satisfy a request has to be done by iterating over them, which
+can be expensive. To mitigate that, requests should be written so that "simple"
+checks are executed first to prune the search tree early.
+
+## Design Details
+
+The semantic model is hosted in
+`k/k/staging/src/k8s.io/dynamic-resource-allocation`:
+
+- `semantic/modeswitch`: the implementation of the builtin controller
+- `apis/modeswitch`: the versioned type(s)
+- `internal/apis/modeswitch`: the internal type for conversion and for use
+  by the implementation
+
+The only modification elsewhere is registering the builtin controller in the
+dynamic resource scheduler plugin if (and only if) the model is enabled.
+
+The semantic model uses a label selector for selecting instances and modes that
+is more capable than `metav1.LabelSelector`. For example, semantic version
+comparison and nested expressions are supported. This label selector is defined
+in its own packages and can be reused by future semantic models, if desired:
+
+- `semantic/labels`: the implementation of the label selector
+- `apis/labels`: the versioned type(s)
+- `internal/apis/labels`: the internal type for conversion and for use
+  by the implementation
+
+### Label selector
+
+The label selector is matched against a set of label name and value pairs. In
+the mode switching model, this set is the combination of labels specified for
+the instance and the labels specified for each mode, with labels of a mode
+overwriting the ones of the instance.
```

Comment:
We can also make the label selector available as a stand-alone semantic model. Then it can be combined with #4382.

 This is not in the KEP yet, but could be added.